proposals/SelectorResourceBudgets.md

```diff
@@ -0,0 +1,207 @@
+	- If we make the budget system non-optional, then existing Selector documents may not work.
+	- Or, there might be no special work needed here, if the budget system is entirely optional.
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+- Selectors in the go-ipld-prime implementation should have resource budgeting.
+- Test fixtures should demonstrate what a selection which halts due to a resource budget exhaustion behaves like.
+- The resource budget specification declaration system should be reasonably comprehensible and look like something we can tell administrators of hypothetical services using this system how to configure.
+	- Probably: it should be as simple as _one number_.
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+When developers such as the Filecoin team feel comfortable exposing features using Selectors to users, then this project is a success.
+
+#### Counterpoints & pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- Overcomplicating the budget system could result in usability failure.
+  (Arguably, the current limit systems are this, because they're too granular, which is no substitute for a holistic system.)
+
+- Technical consideration: Beware the "[Billion Laughs](https://en.wikipedia.org/wiki/Billion_laughs)" problem.
+  (This is why this document keeps emphasizing a budget that is holistic and monotonically decreasing.)
+
+- A system that halts but returns insufficient information about why could be frustrating to users,
+  even if it successfully addresses the DoS problem.
+
+- Keep in mind: this proposal only describes implementing this in golang.
+  We do not currently have a javascript Selectors implementation (and creating one is a larger task).
+  This is not a problem per se; it's just something to remember when considering what can be immediately built upon this work.
+
+#### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+One: See remarks about budgeting in the Assumptions & Hypothesis section.
+Some system of resource currency could be associated with this problem as part of the solution.
+(This doesn't necessarily remove the need for engineering work on the Selector system to support it, though,
+which means this should probably be considered a stretch goal or future work rather than an alternative.)
+
+Two: It's possible to work around this in some cases by building APIs around selectors,
+but then only accept a known, pre-specified set of selectors.
+(If I understand correctly, this is how several pieces of Filecoin currently around around this issue.)
+This is not a general workaround, though, and ruins most of the point of Selectors -- they're *supposed* to be user-specifiable.
```

Comment:
I don't know the specific, usecase but I assume a gimped selector syntax could still be useful potentially for some users depending on their needs.

Reply:
Then we would need to invent such a gimped syntax.

That's probably _harder_ than planning and implementing budgets within the current syntax.

My general experience with this topic is: you cannot, no matter how big of a gavel you wave and how energetically you wave it, convince people to stop asking for features that would make a system accidentally turing-complete (and thus an unbounded DoS vector).  (This is doubly true when it comes to tree or graph processing, which, in a sudden flash hindsight that only occurs to me fully now, probably ought not be a surprise.)  Therefore: monotonically decreasing budgets, often aka "gas", is the only real way to unambiguously communicate the problem, and thus the only real practical way to solve it.

Reply:
The other route is "invent a gimped syntax ***and*** a compiler that verifies its non-TC-ness" -- and that's possible; that's what eBPF is, if I understand correctly -- but it's a _huge_ amount of work...

... and I sorta wouldn't bet very favorably on if that approach would work for tree/DAG processing scenarios anyway.  I'd rather bet money that it would end up with people wanting to apply the eBPF-like thing repeatedly on every block they visit.

Which would get us back to approximately the same problem with Selectors right now: since those things would "restart" their budget on every block visited, we'd need some... bigger, holistic, monotonically decreasing budget.